File: src/main/java/fi/finlit/edith/ui/pages/NoteSearch.java
/*
 * Copyright (c) 2009 Mysema Ltd.
 * All rights reserved.
 *
 */
package fi.finlit.edith.ui.pages;

import static org.apache.commons.lang.StringUtils.isBlank;

import java.util.Collection;
import java.util.HashSet;

import org.apache.tapestry5.Asset;
import org.apache.tapestry5.EventContext;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.Environmental;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.Path;
import org.apache.tapestry5.annotations.Property;
import org.apache.tapestry5.grid.GridDataSource;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.ioc.annotations.Symbol;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

import com.mysema.tapestry.core.Context;

import fi.finlit.edith.EDITH;
import fi.finlit.edith.domain.Concept;
import fi.finlit.edith.domain.Note;
import fi.finlit.edith.ui.services.DocumentRepository;
import fi.finlit.edith.ui.services.NoteDao;
import fi.finlit.edith.ui.services.PrimaryKeyEncoder;

@SuppressWarnings("unused")
@Import(library = { "classpath:js/jquery-1.4.1.js", "deleteDialog.js" })
public class NoteSearch {

    @Property
    private String searchTerm;

    @Property
    private String editMode;

    private Context context;

    @Property
    private GridDataSource notes;

    @Property
    private Note note;

    @Inject
    private NoteDao noteRepository;

    @Inject
    private DocumentRepository documentRepository;

    @Property
    private PrimaryKeyEncoder<Note> encoder;

    @Inject
    @Path("NoteSearch.css")
    private Asset stylesheet;

    @Environmental
    private JavaScriptSupport support;

    private Collection<Note> selectedNotes;

    private Collection<String> orphanNoteIds;

    private boolean removeSelected;
    
    @Inject
    @Symbol(EDITH.EXTENDED_TERM)
    @Property
    private boolean slsMode;

    @AfterRender
    void addStylesheet() {
        // This is needed to have the page specific style sheet after
        // other css includes
//        support.addStylesheetLink(stylesheet, null);
        support.importStylesheet(stylesheet);
    }

    void onActionFromCancel() {
        context = new Context(searchTerm);
    }

    void onActionFromToggleEdit() {
        context = new Context(searchTerm, "edit");
    }

    void onSelectedFromRemoveSelected() {
        removeSelected = true;
    }

    void onActivate(EventContext ctx) {
        // TODO : fetch only orpahNote ids for the elements displayed in this window
        orphanNoteIds = noteRepository.getOrphanIds();

        if (selectedNotes == null){
            selectedNotes = new HashSet<Note>();
        }
        if (ctx.getCount() >= 1) {
            searchTerm = ctx.get(String.class, 0);
        }
        if (ctx.getCount() >= 2) {
            editMode = ctx.get(String.class, 1);
        }
        context = new Context(ctx);
    }

    Object onPassivate() {
        return context == null ? null : context.toArray();
    }

    void onPrepare() {
        encoder = new PrimaryKeyEncoder<Note>(noteRepository);
    }

    void onSuccessFromEdit() {
        if (removeSelected){
            noteRepository.removeNotes(selectedNotes);

        }else{
          //getting all values from encoder
            for(Note editedNote : encoder.getAllValues() ){
              //If we get a not empty value for a field,
              //it means it has been edited
              //We must refetch the actual document note
              if (!isBlank(editedNote.getLemma())) {
                  Note currentNote = noteRepository.getById(editedNote.getId());
                  currentNote.setLemma(editedNote.getLemma());
                  noteRepository.save(editedNote);
              }
            }
        }

        context = new Context(searchTerm);
    }

    void onSuccessFromSearch() {
        context = new Context(searchTerm);
    }

    void setupRender() {
        notes = noteRepository.queryNotes(searchTerm == null ? "*" : searchTerm);
    }

    public boolean isNoteSelected() {
        return selectedNotes.contains(note);
    }

    public void setNoteSelected(boolean selected) {
        if (selected) {
            selectedNotes.add(note);
        } else {
            selectedNotes.remove(note);
        }
    }

    public boolean isNotesNotEmpty(){
        return notes.getAvailableRows() > 0;
    }

    public boolean isOrphanNote(){
        return orphanNoteIds.contains(note.getId());
    }

    public Concept getConcept() {
        return note.getConcept(slsMode);
    }
}

File: src/main/scala/fi/finlit/edith/ui/pages/NoteSearch.scala
package fi.finlit.edith.ui.pages

import org.apache.commons.lang.StringUtils.isBlank
import java.util.Collection
import java.util.HashSet
import org.apache.tapestry5.Asset
import org.apache.tapestry5.EventContext
import org.apache.tapestry5.annotations.AfterRender
import org.apache.tapestry5.annotations.Environmental
import org.apache.tapestry5.annotations.Import
import org.apache.tapestry5.annotations.Path
import org.apache.tapestry5.annotations.Property
import org.apache.tapestry5.grid.GridDataSource
import org.apache.tapestry5.ioc.annotations.Inject
import org.apache.tapestry5.ioc.annotations.Symbol
import org.apache.tapestry5.services.javascript.JavaScriptSupport
import com.mysema.tapestry.core.Context
import fi.finlit.edith.EDITH
import fi.finlit.edith.sql.domain.Note
import fi.finlit.edith.ui.services.DocumentDao
import fi.finlit.edith.ui.services.NoteDao
import fi.finlit.edith.ui.services.SqlPrimaryKeyEncoder
//remove if not needed
import scala.collection.JavaConversions._

@Import(library = ("classpath:js/jquery-1.4.1.js", "deleteDialog.js"))
class NoteSearch {

  @Property
  private var searchTerm: String = _

  @Property
  private var editMode: String = _

  private var context: Context = _

  @Property
  private var notes: GridDataSource = _

  @Property
  private var note: Note = _

  @Inject
  private var noteDao: NoteDao = _

  @Inject
  private var documentDao: DocumentDao = _

  @Property
  private var encoder: SqlPrimaryKeyEncoder[Note] = _

  @Inject
  @Path("NoteSearch.css")
  private var stylesheet: Asset = _

  @Environmental
  private var support: JavaScriptSupport = _

  private var selectedNotes: Collection[Note] = _

  private var orphanNoteIds: Collection[Long] = _

  private var removeSelected: Boolean = _

  @Inject
  @Symbol(EDITH.EXTENDED_TERM)
  @Property
  private var slsMode: Boolean = _

  @AfterRender
  def addStylesheet() {
    support.importStylesheet(stylesheet)
  }

  def onActionFromCancel() {
    context = new Context(searchTerm)
  }

  def onActionFromToggleEdit() {
    context = new Context(searchTerm, "edit")
  }

  def onSelectedFromRemoveSelected() {
    removeSelected = true
  }

  def onActivate(ctx: EventContext) {
    orphanNoteIds = noteDao.getOrphanIds
    if (selectedNotes == null) {
      selectedNotes = new HashSet[Note]()
    }
    if (ctx.getCount >= 1) {
      searchTerm = ctx.get(classOf[String], 0)
    }
    if (ctx.getCount >= 2) {
      editMode = ctx.get(classOf[String], 1)
    }
    context = new Context(ctx)
  }

  def onPassivate(): AnyRef = {
    if (context == null) null else context.toArray()
  }

  def onPrepare() {
    encoder = new SqlPrimaryKeyEncoder[Note](noteDao)
  }

  def onSuccessFromEdit() {
    if (removeSelected) {
      noteDao.removeNotes(selectedNotes)
    } else {
      for (editedNote <- encoder.getAllValues if !isBlank(editedNote.getLemma)) {
        var currentNote = noteDao.getById(editedNote.getId)
        currentNote.setLemma(editedNote.getLemma)
        noteDao.save(editedNote)
      }
    }
    context = new Context(searchTerm)
  }

  def onSuccessFromSearch() {
    context = new Context(searchTerm)
  }

  def setupRender() {
    notes = noteDao.queryNotes(if (searchTerm == null) "*" else searchTerm)
  }

  def isNoteSelected(): Boolean = selectedNotes.contains(note)

  def setNoteSelected(selected: Boolean) {
    if (selected) {
      selectedNotes.add(note)
    } else {
      selectedNotes.remove(note)
    }
  }

  def isNotesNotEmpty(): Boolean = notes.getAvailableRows > 0

  def isOrphanNote(): Boolean = orphanNoteIds.contains(note.getId)
}
